rclpy/src/rclpy/_rclpy.c

```diff
@@ -95,28 +97,31 @@ rclpy_create_guard_condition(PyObject * Py_UNUSED(self), PyObject * Py_UNUSED(ar
   rcl_guard_condition_t * gc =
     (rcl_guard_condition_t *)PyMem_Malloc(sizeof(rcl_guard_condition_t));
   *gc = rcl_get_zero_initialized_guard_condition();
+  PyObject * pygc = PyCapsule_New(gc, "rcl_guard_condition_t", NULL);
   rcl_guard_condition_options_t gc_options = rcl_guard_condition_get_default_options();
 
   rcl_ret_t ret = rcl_guard_condition_init(gc, gc_options);
   if (ret != RCL_RET_OK) {
     PyErr_Format(PyExc_RuntimeError,
       "Failed to create guard_condition: %s", rcl_get_error_string_safe());
+    Py_DECREF(pygc);
     rcl_reset_error();
     return NULL;
   }
   PyObject * pylist = PyList_New(2);
 
-  PyList_SET_ITEM(pylist, 0, PyCapsule_New(gc, NULL, NULL));
+  PyList_SET_ITEM(pylist, 0, pygc);
   PyList_SET_ITEM(pylist, 1, PyLong_FromUnsignedLongLong((uint64_t)&gc->impl));
 
   return pylist;
 }
 
 /// Trigger a general purpose guard condition
 /**
+ * Raises ValueError if the guard condition is not a capsule of the correcttype
+ * Raises RuntimeError if the guard condition could not be triggered
```

Comment:
Several other places do raise runtime error on this type of failure, maybe we should take advantage of this PR to update them all

Reply:
I updated all the ones I found in 451a072e372eecc46e07096f8fdcb2278051bd54